numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -309,6 +314,79 @@ static int
 
 /**end repeat**/
 
+#ifdef HAVE_STRTOLD_L
+static NPY_INLINE npy_longdouble string_to_long_double(PyObject*op)
+{
+    char *s;
+    char *end;
+    npy_longdouble temp;
+    PyObject* b;
+
+    /* This weird structure is to make the error handling work:
+     * if it's not a string, we want to fall back to other
+     * approaches. In python 2, that happens when we try to
+     * convert the string to a C string. In python 3, we have
+     * to use a two-step conversion process, via bytes; the
+     * not-a-string situation gets triggered on conversion to 
+     * bytes, but we want to make sure bytes objects get 
+     * handled too, because python accepts bytes as input to
+     * float() (why?).
+     */
+    b = PyUnicode_AsUTF8String(op);
```

Comment:
I'm worried about duck typing. Basically if the input is string-like enough, I want it to get turned into a string. Unfortunately on python 2 non-unicode strings go through two conversions, so maybe a check would be better.

And if the string isn't ASCII it's not a valid number, but if this fails (by UnicodeError) we fall through to conversion via a python float, whereas if this succeeds we report an inability to convert from strtold. So UTF-8 avoids conversion failures here.